### PR TITLE
Relevad RTD Module : initial release

### PR DESCRIPTION
**Pull request by aenrel**
## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes

- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

This pull request implements Relevad RTD module.
Please review it
Thanks a lot!

### CODE REVIEW

File: modules/relevadRtdProvider.js [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+            case 'smart':
+            case 'smartadserver':
+              let target = [].concat(bid.params?.target ? [bid.params.target] : []);
+              processBidder('smart', function() {
+                target = target.concat(relevadList.map(entry => { return 'relevad_rtd=' + entry; }));
+                deepSetValue(bid, 'params.target', target.join(';'));
+              });
+              break;
+
+            case 'ix':
```
Comment (patmmccann) [inline modules/relevadRtdProvider.js:314]: our goal is to eliminate all this case handling in rtd adapters and move them into bidders. You can pr all the adapters that require special handling of the standard object.

**Review by patmmccann — CHANGES_REQUESTED** [review]
pls move bid adapter handling into bid adapters

File: modules/relevadRtdProvider.js (hunk 1 above)
Comment (aenrel) [reply modules/relevadRtdProvider.js:314]: Hi,
I took the case handling code almost literally from the already approved RTD provider [sirdataRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/sirdataRtdProvider.js#L230), and my code does exactly the same case handling.
Many other approved RTD providers have very similar case handling code in order to accomodate the non-standard server-side of various bidders, for example:

- [sirdataRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/sirdataRtdProvider.js#L230)
- [permutiveRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/permutiveRtdProvider.js#L253)
- [weboramaRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/weboramaRtdProvider.js#L690)
- [captifyRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/captifyRtdProvider.js#L43)
- [1plusXRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/1plusXRtdProvider.js#L246)

I don’t think I have the authority to change other companies  (appnexus, rubicon etc) BidAdapter code.
Simply removing all case handling in my module would cripple it’ functionality vs. the others, and I’d like to keep it competitive.

Considering this, could you please let  the case handling stay in my code?
Please  advise how to proceed

Thank you very much!

**Comment by aenrel** [comment]
Hi, I'd like to find out whether I can keep bid adapter case handling in my code based on these considerations.
Thank you very much!

I took the case handling code almost literally from the already approved RTD provider [sirdataRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/sirdataRtdProvider.js#L230), and my code does exactly the same case handling.
Many other approved RTD providers have very similar case handling code in order to accomodate the non-standard server-side of various bidders, for example:

[sirdataRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/sirdataRtdProvider.js#L230)
[permutiveRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/permutiveRtdProvider.js#L253)
[weboramaRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/weboramaRtdProvider.js#L690)
[captifyRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/captifyRtdProvider.js#L43)
[1plusXRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/1plusXRtdProvider.js#L246)
I don’t think I have the authority to change other companies (appnexus, rubicon etc) BidAdapter code.
Simply removing all case handling in my module would cripple it’ functionality vs. the others, and I’d like to keep it competitive.

Considering this, could you please let the case handling stay in my code?
Please advise how to proceed

File: modules/relevadRtdProvider.js (hunk 1 above)
Comment (aenrel) [reply modules/relevadRtdProvider.js]: Addressed your request w.r.t. removing bidder-specific handling' with this commit:
https://github.com/prebid/Prebid.js/pull/9454/commits/6967feb082d477d1f956a505ea8a19d21d8e23b2

Please review it
Thank you

**Comment by aenrel** [comment]
Addressed your request w.r.t. removing bidder-specific handling' with this commit:
https://github.com/prebid/Prebid.js/commit/6967feb082d477d1f956a505ea8a19d21d8e23b2

Please review it
Thank you

**Comment by aenrel** [comment]
@ncolletti ping
Could you please review this pull request
Thank you!

File: modules/relevadRtdProvider.js [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+ *
+ * @param      {dictionary}  ortb2     The gloabl ORTB structure
+ * @param      {dictionary}  rtdData   Rtd segments and categories
+ */
+export function setGlobalOrtb2(ortb2, rtdData) {
+  try {
+    let addOrtb2 = composeOrtb2Data(rtdData, 'site');
+    !isEmpty(addOrtb2) && mergeDeep(ortb2, addOrtb2);
+  } catch (e) {
+    logError(e)
```
Comment (ncolletti) [inline modules/relevadRtdProvider.js:92]: would be helpful to include the submodule name prefix

File: modules/relevadRtdProvider.js [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+    let addOrtb2 = composeOrtb2Data(rtdData, 'site');
+    !isEmpty(rtdData.segments) && deepSetValue(addOrtb2, 'user.ext.data.relevad_rtd', rtdData.segments);
+    !isEmpty(rtdData.categories?.sectioncat) && deepSetValue(addOrtb2, 'site.ext.data.relevad_rtd', rtdData.categories.sectioncat);
+    if (isEmpty(addOrtb2)) {
+      return;
+    }
+    bidderOrtbFragment[bidder] = bidderOrtbFragment[bidder] || {};
+    mergeDeep(bidderOrtbFragment[bidder], addOrtb2);
+  } catch (e) {
+    logError(e)
```
Comment (ncolletti) [inline modules/relevadRtdProvider.js:144]: include submodule name prefix here and other areas of the PR where it's missing, thank you

File: modules/relevadRtdProvider.js [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+    gdprc: encodeURIComponent(userConsent?.gdpr?.consentString || ''),
+  }
+  if (!config.dryrun) {
+    data.page = serverData?.page || config?.params?.actualUrl || getRefererInfo().page || '';
+  }
+
+  sendBids(data, config);
+
+  serverData = {};
+  return data;
```
Comment (ncolletti) [inline modules/relevadRtdProvider.js:353]: I don't believe `onAuctionEndEvent` expects a return value

File: modules/relevadRtdProvider.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+Module Name: Relevad Rtd Provider
+Module Type: Rtd Provider
+Maintainer: anna@relevad.com
+
+# Description
+
+Relevad is a contextual semantic analytics company. Our privacy-first, cookieless contextual categorization, segmentation, and keyword generation platform is designed to help publishers and advertisers optimize targeting and increase ad inventory yield.
+
+This real-time data processing module provides quality contextual IAB categories and segments to prebid.js auction bidders. Every category and segment comes with a score measuring its relevance to the publisher’s web page. Publishers may configure their preferred minimum score setting with the module parameter “minscore”. Relevad service does not use browser cookies and is fully GDPR compliant.
```
Comment (ncolletti) [inline modules/relevadRtdProvider.md:11]: Might be helpful in this section to clarify that the received cats and segs are being applied in the bid request, as well as give a format to expect

**Review by ncolletti — COMMENTED** [review]
Looks good, few small comments to address. Also, before we merge we'd like to see an update to your example html. The parameters should include all required properties with valid values to make a request to your endpoint. We should see how your module appends segments in the request and adds targeting.

Also, could you add documentation in your md to give a sense of what to expect from the gpt targeting data that is added. Examples of format would be helpful.

File: modules/relevadRtdProvider.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+| minscore | Integer | Bidder-specific minimum categorization relevancy score (0, 100) | Optional, defaults to global minscore above. |
+
+If you do not have your own `partnerid, publisherid, apikey` please reach out to [anna@relevad.com](mailto:anna@relevad,com).
+
+### Testing
+
+To view an example of the on page setup required:
+
+```bash
+gulp serve-fast --modules=rtdModule,relevadRtdProvider,appnexusBidAdapter
```
[the remark this answers is not in the document]
Comment (aenrel) [reply modules/relevadRtdProvider.md]: Removed, thank you!

File: modules/relevadRtdProvider.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+#### Bidder-specific configuration. Every bidder may have these configuration parameters
+
+| Name  |Type | Description   | Notes  |
+| :------------ | :------------ | :------------ |:------------ |
+| bidder | String | Bidder name | Mandatory. Example: “appnexus” |
+| adUnitCodes | Array of Strings | List of specific AdUnit codes you with to target | Optional. If empty or absent, all ad units are targeted. |
+| minscore | Integer | Bidder-specific minimum categorization relevancy score (0, 100) | Optional, defaults to global minscore above. |
+
+If you do not have your own `partnerid, publisherid, apikey` please reach out to [anna@relevad.com](mailto:anna@relevad,com).
```
[the remark this answers is not in the document]
Comment (aenrel) [reply modules/relevadRtdProvider.md]: Fixed, thank you!

File: modules/relevadRtdProvider.js (hunk 4 above)
Comment (aenrel) [reply modules/relevadRtdProvider.js]: Fixed, thank you!

File: modules/relevadRtdProvider.md (hunk 5 above)
Comment (aenrel) [reply modules/relevadRtdProvider.md]: Added description
Thank you!

**Review by ncolletti — APPROVED** [review]
looks good! as a follow up to this initial release, we generally recommend partners include live parameters in their example html pages. this users to see how requests and responses are handled in an ideal scenario for your module.

**Comment by ncolletti** [comment]
@patmmccann your changes were addressed, ready for you to take a look

**Comment by patmmccann** [comment]
LGTM

**Review by patmmccann — APPROVED** [review]

**Pull request merged**